library/proc_macro/src/bridge/server.rs

```diff
@@ -5,22 +5,24 @@ use super::*;
 // FIXME(eddyb) generate the definition of `HandleStore` in `server.rs`.
 use super::client::HandleStore;
 
-/// Declare an associated item of one of the traits below, optionally
-/// adjusting it (i.e., adding bounds to types and default bodies to methods).
-macro_rules! associated_item {
-    (type FreeFunctions) => (type FreeFunctions: 'static;);
-    (type TokenStream) => (type TokenStream: 'static + Clone;);
-    (type TokenStreamBuilder) => (type TokenStreamBuilder: 'static;);
-    (type TokenStreamIter) => (type TokenStreamIter: 'static + Clone;);
-    (type Group) => (type Group: 'static + Clone;);
-    (type Punct) => (type Punct: 'static + Copy + Eq + Hash;);
-    (type Ident) => (type Ident: 'static + Copy + Eq + Hash;);
-    (type Literal) => (type Literal: 'static + Clone;);
-    (type SourceFile) => (type SourceFile: 'static + Clone;);
-    (type MultiSpan) => (type MultiSpan: 'static;);
-    (type Diagnostic) => (type Diagnostic: 'static;);
-    (type Span) => (type Span: 'static + Copy + Eq + Hash;);
+pub trait Types {
+    type FreeFunctions: 'static;
+    type TokenStream: 'static + Clone;
+    type TokenStreamBuilder: 'static;
+    type TokenStreamIter: 'static + Clone;
+    type Group: 'static + Clone;
+    type Punct: 'static + Copy + Eq + Hash;
+    type Ident: 'static + Copy + Eq + Hash;
+    type Literal: 'static + Clone;
+    type SourceFile: 'static + Clone;
+    type MultiSpan: 'static;
+    type Diagnostic: 'static;
+    type Span: 'static + Copy + Eq + Hash;
+}
```

Comment:
It would really be nice if this info was in `with_api!` instead, it just sucks that it would have to look like `: (...);` instead of `: ...;` to be easily digestible by proc macros.

Reply:
I thought about adding these bounds to `with_api_types!` so that this trait could be a `with_api_types` invocations and decided against it.

I understand the appeal of DRY and the platonic goal of encoding all of the API information in a single macro so that API changes (e.g. adding new types and/or methods) only requires modifying a single code location. But I don't think that's so important. If adding a new type and/or method requires modifying two or three places (which the compiler will point out for you) it's not that bad.

What *is* bad, and I have had to struggle with for a couple of days now, is how hard it is to read really macro-heavy code. I have found this bridge code hard to read. I find it hard to do macro substitution in my head. I've written out parts of the expansions of several macro invocations in a separate text file just to see what they look like. I'm contemplating putting such example expansions into comments to help future readers of the code (including me).